lib/matplotlib/category.py

```diff
@@ -184,10 +207,17 @@ def update(self, data):
         data = np.atleast_1d(np.array(data, dtype=object))
 
         for val in OrderedDict.fromkeys(data):
+            # OrderedDict just iterates over unique values in data.
             if not isinstance(val, (str, bytes)):
                 raise TypeError("{val!r} is not a string".format(val=val))
             if val not in self._mapping:
                 self._mapping[val] = next(self._counter)
+        # check if we can convert all strings to number or date...
+        if self._strs_are_convertible(data):
+            _log.info('using category units to plot a list of '
```

Comment:
Use self.__mapping.keys() instead of data so you at least don't have to double iterate over everything.

Reply:
The keys will include keys from previous calls to the converter, which may indeed be valid.  So thats different than the check here, which tests only the most recent data that are sent to the converter.  Is that what you want?

Reply:
Hmm...I'm just really not liking the second loop, but a second call to unique under the hood would be doing the same thing. :/

Reply:
OK, someone told me to take it out of the loop, but... new version has it in the loop and only calls the check so long as the strings are convertable to numbers.  (i.e. stops checking as soon as a non-number is found).